File: frontend/src/services/downloads/download-collection-image.ts
import { collectionItemInfo } from "@/helper/item-info";
import apiClient from "@/services/api-client";
import { ReturnErrorMessage } from "@/services/api-error-return";

import { log } from "@/lib/logger";

import type { APIResponse } from "@/types/api/api-response";
import type { CollectionItem } from "@/types/media-and-posters/collection-item";
import type { CollectionItemImageFile } from "@/types/media-and-posters/sets";
import { TYPE_DOWNLOAD_COLLECTION_IMAGE_TYPE_OPTIONS } from "@/types/ui-options";

export interface DownloadCollectionImage_Request {
  collection_item: CollectionItem;
  image_file: CollectionItemImageFile;
}

export interface DownloadCollectionImage_Response {
  result: string;
}

export const DownloadImageFileForCollectionItem = async (
  imageType: TYPE_DOWNLOAD_COLLECTION_IMAGE_TYPE_OPTIONS,
  collectionItem: CollectionItem,
  imageFile: CollectionItemImageFile
): Promise<APIResponse<DownloadCollectionImage_Response>> => {
  log(
    "INFO",
    "API - Media Server",
    "Download and Update Collection Image",
    `Downloading ${imageType} image and updating ${collectionItemInfo(collectionItem)}`
  );
  try {
    const req: DownloadCollectionImage_Request = {
      collection_item: collectionItem,
      image_file: imageFile,
    };
    const response = await apiClient.post<APIResponse<DownloadCollectionImage_Response>>(
      `/download/image/collection`,
      req
    );
    if (response.data.status === "error") {
      throw new Error(
        response.data.error?.message || `Unknown error downloading ${imageType} image and updating media server`
      );
    } else {
      log(
        "INFO",
        "API - Media Server",
        "Download and Update Collection Image",
        `Downloaded ${imageType} image and updated ${collectionItemInfo(collectionItem)}`,
        response.data
      );
    }
    return response.data;
  } catch (error) {
    log(
      "ERROR",
      "API - Media Server",
      "Download and Update Collection Image",
      `Failed to download ${imageType} image and update ${collectionItemInfo(collectionItem)}
			${error instanceof Error ? error.message : "Unknown error"}`,
      error
    );
    return ReturnErrorMessage<DownloadCollectionImage_Response>(error);
  }
};

File: backend/routing/download/download_collection_image.go
package routes_download

import (
	"aura/logging"
	"aura/mediaserver"
	"aura/models"
	"aura/utils/httpx"
	"net/http"
)

type DownloadCollectionImage_Request struct {
	ImageFile      models.ImageFile      `json:"image_file"`
	CollectionItem models.CollectionItem `json:"collection_item"`
}

type DownloadCollectionImage_Response struct {
	Result         string `json:"result"`
	ArtworkVersion int64  `json:"artwork_version,omitempty"`
}

// DownloadImageFileForCollectionItem godoc
// @Summary      Download Collection Image
// @Description  Download and apply an image for a Collection Item in the media server.
// @Tags         Download
// @Accept       json
// @Produce      json
// @Param        req  body      DownloadCollectionImage_Request  true  "Download Collection Image Request"
// @Security 	 BearerAuth
// @Failure      401  {object}  httpx.UnauthorizedResponse "Unauthorized (only when Auth.Enabled=true)"
// @Success      200           {object}  httpx.JSONResponse{data=DownloadCollectionImage_Response}
// @Failure      500           {object}  httpx.JSONResponse "Internal Server Error"
// @Router       /api/download/image/collection [post]
func DownloadImageFileForCollectionItem(w http.ResponseWriter, r *http.Request) {
	ctx, ld := logging.CreateLoggingContext(r.Context(), r.URL.Path)
	logAction := ld.AddAction("Download Collection Image", logging.LevelInfo)
	ctx = logging.WithCurrentAction(ctx, logAction)
	var req DownloadCollectionImage_Request
	var response DownloadCollectionImage_Response

	// Parse the request body
	Err := httpx.DecodeRequestBodyToJSON(ctx, r.Body, &req, "Download Collection Image - Decode Request Body")
	if Err.Message != "" {
		httpx.SendResponse(w, ld, response)
		return
	}

	actionValidate := logAction.AddSubAction("Validate Request Body Fields", logging.LevelDebug)
	// Make sure that the collectionItem has the following fields set
	// 1. CollectionItem.RatingKey
	// 2. CollectionItem.LibraryTitle
	if req.CollectionItem.RatingKey == "" || req.CollectionItem.LibraryTitle == "" {
		actionValidate.SetError("Invalid Collection Item Data", "RatingKey and LibraryTitle are required in Collection Item", map[string]any{
			"rating_key":    req.CollectionItem.RatingKey,
			"library_title": req.CollectionItem.LibraryTitle,
		})
		actionValidate.Complete()
		httpx.SendResponse(w, ld, response)
		return
	}

	// Make sure that the ImageFile has the following fields set
	// 1. ImageFile.ID
	// 2. ImageFile.Type
	if req.ImageFile.ID == "" || req.ImageFile.Type == "" {
		actionValidate.SetError("Invalid Image File Data", "ID and Type are required in Image File", map[string]any{
			"id":   req.ImageFile.ID,
			"type": req.ImageFile.Type,
		})
		actionValidate.Complete()
		httpx.SendResponse(w, ld, response)
		return
	}
	actionValidate.Complete()

	// Make the download and apply the image
	Err = mediaserver.ApplyCollectionImage(ctx, &req.CollectionItem, req.ImageFile)
	if Err.Message != "" {
		httpx.SendResponse(w, ld, response)
		return
	}

	response.Result = "ok"
	response.ArtworkVersion = req.CollectionItem.ArtworkVersion
	httpx.SendResponse(w, ld, response)
}
